ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/SourceBuilder.java

```diff
@@ -232,4 +266,116 @@ private static Formats of(final KeyFormat keyFormat, final FormatInfo valueForma
     );
   }
 
+  private static class AddPseudoColumnsToMaterialize<K>
+      implements ValueTransformerWithKeySupplier<K, GenericRow, GenericRow> {
+
+    private final int pseudoColumnVersion;
+
+    AddPseudoColumnsToMaterialize(final int pseudoColumnVersion) {
+      this.pseudoColumnVersion = pseudoColumnVersion;
+    }
+
+    @Override
+    public ValueTransformerWithKey<K, GenericRow, GenericRow> get() {
+      return new ValueTransformerWithKey<K, GenericRow, GenericRow>() {
+        private ProcessorContext processorContext;
+
+        @Override
+        public void init(final ProcessorContext processorContext) {
+          this.processorContext = requireNonNull(processorContext, "processorContext");
+        }
+
+        @Override
+        public GenericRow transform(final K key, final GenericRow row) {
+          if (row == null) {
+            return row;
+          }
+
+          final int numPseudoColumns = SystemColumns
+              .pseudoColumnNames(pseudoColumnVersion).size();
+
+          row.ensureAdditionalCapacity(numPseudoColumns - 1);
```

Comment:
Why -1? This does not look like it is extensible if additional pseudocolumns are added later. It would be better for the logic for which pseudocolumns need to be materialized to be centralized into a single place, such as where the pseudocolumns are defined.

Reply:
I could add logic into `SystemColumns` which divides the set corresponding to a given `pseudoColumnVersion` into columns which may need to be materialized for access in downstream processor contexts vs those which are not. Is this what you had in mind?

Reply:
That's conceptually what I had in mind but I suspect there are probably cleaner ways to implement it. For example, you could have the SystemColumns class store of list of pseudocolumn objects/structs where each one has a name and a boolean that specifies whether the column should be materialized or not.